### PR TITLE
Add test cases and edge cases for Book and Patron

### PR DESCRIPTION
**Pull request by yl-ang**
- Add JsonSerializableLibTask test cases to consider many patrons to load from the Json data file.
- Add edge cases for PatronName and PatronId.
- Add edge cases for BookName.
- Add edge cases for DeleteBookCommandParser and DeletePatronCommandParser.

### CODE REVIEW

File: src/test/java/seedu/address/model/patron/IdTest.java [hunk 1]
```diff
@@ -31,10 +31,13 @@ public void isValidId() {
...
 
-
         // valid id
         assertTrue(Id.isValidId("A0123456H"));
         assertTrue(Id.isValidId("A0123457K"));
         assertTrue(Id.isValidId("A0123458L"));
+
+        // valid id but will need to check again in the future
+        assertTrue(Id.isValidId("ABCDEFGHI")); // starts with A and only contains letters
+        assertTrue(Id.isValidId("A00000000")); // starts with A followed by 8 digits
```
Comment (yuqitanyq) [inline src/test/java/seedu/address/model/patron/IdTest.java:41]: Great job on adding in edge cases here as it helps in stress testing the patron class. 👍

File: src/test/java/seedu/address/testutil/ManyPatrons.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        return ab;
+    }
+
+    public static List<Patron> getManyPatrons() {
+        return new ArrayList<>(Arrays.asList(ALICE_PAULINE, BENSON_MEIER, CARL_KURZ, DANIEL_MEIER, ELLE_MEYER,
+                FIONA_KUNZ, GEORGE_BEST, AMY, ROBERT, JEAN, THOMAS, TOMMY_GOH, TOMMY_LIM, ALICE_MAKER, BENSON_TOH,
+                CARL_KARL, DANIEL_NG, MASHMALLOW, FIONA_RAIN, GEORGER_WONG, AMY_LOW, ROBBERT_HOOK, JEAN_CHUA,
+                THOMAS_TAN));
+    }
+}
```
Comment (yuqitanyq) [inline src/test/java/seedu/address/testutil/ManyPatrons.java:104]: Great job on adding a comprehensive set of 30 patrons here for testing. 👍

File: src/test/java/seedu/address/logic/parser/book/DeleteBookCommandParserTest.java [hunk 3]
```diff
@@ -21,5 +21,8 @@ public void parse_validArgs_returnsDeleteCommand() {
     public void parse_invalidArgs_throwsParseException() {
         assertParseFailure(parser, "a", String.format(MESSAGE_INVALID_COMMAND_FORMAT,
                 DeleteBookCommand.MESSAGE_USAGE));
+
+        assertParseFailure(parser, "-999", String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                DeleteBookCommand.MESSAGE_USAGE));
```
Comment (yuqitanyq) [inline src/test/java/seedu/address/logic/parser/book/DeleteBookCommandParserTest.java:26]: I like the crafting of the edge case here and the attention to detail on ensuring that the classes work as intended! 👍

**Review by yuqitanyq — APPROVED** [review]
LGTM! 👍 Great job on crafting out test cases and edge cases for Book and Patron. I like how thoughtful some of the cases are as it really tests the classes comprehensively. Keep it up!

File: src/test/java/seedu/address/logic/parser/book/DeleteBookCommandParserTest.java (hunk 3 above)
Comment (bingsen0806) [inline src/test/java/seedu/address/logic/parser/book/DeleteBookCommandParserTest.java:26]: Good use of negative test case!

File: src/test/java/seedu/address/model/book/BookNameTest.java [hunk 4]
```diff
@@ -31,6 +31,12 @@ public void isValidBookName() {
         assertFalse(BookName.isValidBookName("Harry&Potter")); // contains non-alphanumeric characters not : or '
         assertFalse(BookName.isValidBookName(" HarryPotter")); // starts with non-alphanumeric character
 
+        // invalid book name but should be valid books
+        assertFalse(BookName.isValidBookName("Who Moved My Cheese?")); // ends with "?"
+        assertFalse(BookName.isValidBookName("4.50_from_Paddington")); // contains "."
+        assertFalse(BookName.isValidBookName("11/22/63")); // contains "/"
+        assertFalse(BookName.isValidBookName("Winnie-the-Pooh")); // contains "-"
```
Comment (bingsen0806) [inline src/test/java/seedu/address/model/book/BookNameTest.java:38]: These may not be necessary because they are in the same equivalent partition as test case of line 31. It may incur a time cost of changing test cases if we change the regex to accept other characters in the future.

File: src/test/java/seedu/address/model/book/BookTest.java [hunk 5]
```diff
@@ -140,4 +140,20 @@ public void equals() {
         editedHarryPotter = new BookBuilder(HARRY_POTTER).withBookStatus(getSampleBorrowedStatus()).build();
         assertFalse(HARRY_POTTER.equals(editedHarryPotter));
     }
+
+    @Test
+    public void isBookAvailable_bookAvailable_returnsTrue() {
```
Comment (bingsen0806) [inline src/test/java/seedu/address/model/book/BookTest.java:145]: Good attempt for coverage by adding both positive and negative test cases.

**Review by bingsen0806 — APPROVED** [review]
LGTM! Good attempt in adding test cases. However, I do not understand the need to ManyPatrons test. Is the test supposed to cover some cases in some equivalent partition or boundaries that were previously not covered?
